### ISSUE TITLE
Outdated API URL

### ISSUE THREAD

**Issue by hcartagena** [ychoi-kr/omegat-plugin-openai-translate#2]
Ser, your plugin looks very promising.

I have a few requests for you, though, hoping to make it even more useful.

If possible, please:

- Add the ability to specify the desired model. For example: gpt-3.5-turbo, gpt-3.5-turbo-16k, gpt-4.

- Add the ability to specify roles (i.e., system, user, assisstant). The 'system' role is particularly useful, as explained [here](https://platform.openai.com/docs/guides/gpt/chat-completions-api). If implemented, users could send "custom instructions" along with the text that needs translation in order to "fine tune" the model's response. This is currently possible with [ChatGPT](https://openai.com/blog/custom-instructions-for-chatgpt), OpenAI's [Playground](https://platform.openai.com/playground?mode=chat) and with [OpenAI Translator](https://community.rws.com/product-groups/trados-portfolio/rws-appstore/w/wiki/6651/openai-translator) for RWS Trados.

- Make a .jar file available for users who don't know how to compile Java files.

Also, as mentioned in the title of this message, the current API URL specified in the Java file is outdated. The current URL is https://api.openai.com/v1/chat/completions. This is the API endpoint used by the newer models (gpt-4, gpt-3.5-turbo). Perhaps, adding the ability to specify it would also be useful.

Thanks!

**Comment by hcartagena** [comment]
I can't confirm this because I am not a programmer, but the following code may fulfill my request above. Can anyone confirm if it works?

import javax.swing.JOptionPane;
import org.omegat.core.data.SourceTextEntry;
import org.omegat.core.data.TMXEntry;
import org.omegat.gui.issues.*;
import org.omegat.util.Language;
import org.omegat.util.WikiGet;
import org.json.JSONObject;
import java.util.Map;
import java.util.HashMap;

public class OpenAITranslate extends BaseTranslate {
    private static final String API_URL = "https://api.openai.com/v1/chat/completions";
    private static final String API_KEY = System.getProperty("openai.api.key");
    private static final Map<String, String> translationCache = new HashMap<>();
    private static final String MODEL = System.getProperty("openai.model", "gpt-3.5-turbo");
    private static final double TEMPERATURE = Double.parseDouble(System.getProperty("openai.temperature", "1"));
    private static final int MAX_TOKENS = Integer.parseInt(System.getProperty("openai.max_tokens", "1000"));
    private static final double TOP_P = Double.parseDouble(System.getProperty("openai.top_p", "1"));

    @Override
    protected String getPreferenceName() {
        return "allow_openai_translate";
    }

    @Override
    public String getName() {
        if (API_KEY == null) {
            return "OpenAI Translate (API Key Required)";
        } else {
            return "OpenAI Translate";
        }
    }

    @Override
    protected String translate(Language sLang, Language tLang, String text) throws Exception {
        if (API_KEY == null) {
            return "";
        }

        // Ask for user confirmation before translating.
        int confirmResult = JOptionPane.showConfirmDialog(null, "Are you sure you want to translate this segment with OpenAI Translate?", "Translation Confirmation", JOptionPane.YES_NO_OPTION);

        if (confirmResult != JOptionPane.YES_OPTION) {
            // If the user does not confirm, the translation is canceled and the original text is returned.
            return text;
        }

        // Request the user to enter a "system message".
        String systemMessage = JOptionPane.showInputDialog("Please enter special instructions for OpenAI Translate to translate the current segment:");

        if (systemMessage == null || systemMessage.isEmpty()) {
            // If the user does not enter a "system message" or leaves it blank, a default message is used.
            systemMessage = "You are a helpful translation assistant.";
        }

        String lvSourceLang = sLang.getLanguageCode().substring(0, 2).toUpperCase();
        String lvTargetLang = tLang.getLanguageCode().substring(0, 2).toUpperCase();

        String cacheKey = lvSourceLang + "-" + lvTargetLang + "-" + text;
        String cachedResult = translationCache.get(cacheKey);
        if (cachedResult != null) {
            return cachedResult;
        }

        // Use the additional system variables.
        String model = MODEL;
        int max_tokens = MAX_TOKENS;

        String prompt = "Translate the following into " + lvTargetLang + ": " + text;
        String escapedPrompt = prompt.replaceAll("\"", "\\\"").replaceAll("\n", "\\n");

        Map<String, String> headers = new HashMap<>();
        headers.put("Content-Type", "application/json");
        headers.put("Authorization", "Bearer " + API_KEY);

        String body = "{\n" +
                "    \"model\": \"" + model + "\",\n" +
                "    \"messages\": [{\"role\": \"system\", \"content\": \"" + systemMessage + "\"}, {\"role\": \"user\", \"content\": \"" + escapedPrompt + "\"}],\n" +
                "    \"max_tokens\": " + max_tokens + ",\n" +
                "    \"temperature\": " + TEMPERATURE + ",\n" +
                "    \"top_p\": " + TOP_P + "\n" +
                "}";

        String response = WikiGet.postJSON(API_URL, body, headers);
        JSONObject jsonResponse = new JSONObject(response);
        String translatedText = jsonResponse.getJSONArray("choices").getJSONObject(0).getString("message").trim();

        translationCache.put(cacheKey, translatedText);
        return translatedText;
    }
}

**Comment by hcartagena** [comment]
Someone smarter than me could tell if this would actually work...
```
import javax.swing.JOptionPane;
import org.omegat.core.machinetranslators.BaseTranslate;
import org.omegat.core.data.SourceTextEntry;
import org.omegat.core.data.TMXEntry;
import org.omegat.util.Language;
import org.omegat.util.HttpConnectionUtils;
import org.json.*;
import java.util.Map;
import java.util.HashMap;
import java.util.TreeMap;

public class OpenAITranslate extends BaseTranslate {
    private static final String API_URL = "https://api.openai.com/v1/chat/completions";
    private static final String API_KEY = System.getProperty("openai.api.key");
    private static final Map<String, String> translationCache = new HashMap<>();
    private static final String MODEL = System.getProperty("openai.model", "gpt-3.5-turbo");
    private static final double TEMPERATURE = Double.parseDouble(System.getProperty("openai.temperature", "1"));
    private static final int MAX_TOKENS = Integer.parseInt(System.getProperty("openai.max_tokens", "1000"));
    private static final double TOP_P = Double.parseDouble(System.getProperty("openai.top_p", "1"));

    @Override
    protected String getPreferenceName() {
        return "allow_openai_translate";
    }

    @Override
    public String getName() {
        if (API_KEY == null) {
            return "OpenAI Translate (API Key Required)";
        } else {
            return "OpenAI Translate";
        }
    }

    @Override
    protected String translate(Language sLang, Language tLang, String text) throws Exception {
        if (API_KEY == null) {
            return "";
        }

        // Ask for user confirmation before translating.
        int confirmResult = JOptionPane.showConfirmDialog(null, "Are you sure you want to translate this segment with OpenAI Translate?", "Translation Confirmation", JOptionPane.YES_NO_OPTION);

        if (confirmResult != JOptionPane.YES_OPTION) {
            // If the user does not confirm, the translation is canceled and the original text is returned.
            return text;
        }

        // Request the user to enter a "system message".
        String systemMessage = JOptionPane.showInputDialog("Please enter special instructions for OpenAI Translate to translate the current segment:");

        if (systemMessage == null || systemMessage.isEmpty()) {
            // If the user does not enter a "system message" or leaves it blank, a default message is used.
            systemMessage = "You are a helpful translation assistant.";
        }

        String lvSourceLang = sLang.getLanguageCode().substring(0, 2).toUpperCase();
        String lvTargetLang = tLang.getLanguageCode().substring(0, 2).toUpperCase();

        String cacheKey = lvSourceLang + "-" + lvTargetLang + "-" + text;
        String cachedResult = translationCache.get(cacheKey);
        if (cachedResult != null) {
            return cachedResult;
        }

        // Use the additional system variables.
        String model = MODEL;
        int max_tokens = MAX_TOKENS;

        String prompt = "Translate the following into " + lvTargetLang + ": " + text;
        String escapedPrompt = prompt.replaceAll("\"", "\\\"").replaceAll("\n", "\\n");

        Map<String, String> headers = new HashMap<>();
        headers.put("Content-Type", "application/json");
        headers.put("Authorization", "Bearer " + API_KEY);

        String body = "{\n" +
                "    \"model\": \"" + model + "\",\n" +
                "    \"messages\": [{\"role\": \"system\", \"content\": \"" + systemMessage + "\"}, {\"role\": \"user\", \"content\": \"" + escapedPrompt + "\"}],\n" +
                "    \"max_tokens\": " + max_tokens + ",\n" +
                "    \"temperature\": " + TEMPERATURE + ",\n" +
                "    \"top_p\": " + TOP_P + "\n" +
                "}";

        String response = HttpConnectionUtils.postJSON(API_URL, body, headers);
        JSONObject jsonResponse = new JSONObject(response);
        String translatedText = jsonResponse.getJSONArray("choices").getJSONObject(0).getString("message").trim();

        translationCache.put(cacheKey, translatedText);
        return translatedText;
    }
}
```

**Comment by ychoi-kr** [comment]
@hcartagena 

Thank you for opening the issue.

I just released v0.1.1 which replacing deprecated text-davinci-003 to gpt-3.5-turbo-instruct. (See https://platform.openai.com/docs/deprecations/instructgpt-models )

By the way, I also tested code which uses Chat Completion API but I doubt it is proper for our need despite the instruct model.

```
package kr.ychoi.otplugin;

import java.util.TreeMap;
import java.util.Map;
import java.util.HashMap;
import org.omegat.core.machinetranslators.BaseTranslate;
import org.omegat.util.Language;
import org.omegat.util.WikiGet;
import org.json.*;

/*
 * OpenAI Translate plugin for OmegaT
 * based on Naver Translate plugin by ParanScreen https://github.com/ParanScreen/omegat-plugin-navertranslate
 * licensed under GNU GPLv2 and modified by ychoi
 */


@SuppressWarnings({"unchecked", "rawtypes"})
public class OpenAITranslate extends BaseTranslate {

    private static final String API_URL = "https://api.openai.com/v1/chat/completions";
    private static final String API_KEY = System.getProperty("openai.api.key");

    private static final Map<String, String> translationCache = new HashMap<>();

    @Override
    protected String getPreferenceName() {
        return "allow_openai_translate";
    }

    public String getName() {
        if (API_KEY == null) {
            return "OpenAI Translate (API Key Required)";
        } else {
            return "OpenAI Translate";
        }
    }

    @Override
    protected String translate(Language sLang, Language tLang, String text) throws Exception {
        if (API_KEY == null) {
            return "";
        }

        String lvSourceLang = sLang.getLanguageCode().substring(0, 2).toUpperCase();
        String lvTargetLang = tLang.getLanguageCode().substring(0, 2).toUpperCase();

        String cacheKey = lvSourceLang + '-' + lvTargetLang + text;
        String cachedResult = translationCache.get(cacheKey);
        if (cachedResult != null) {
            return cachedResult;
        }
        
        String model = "gpt-3.5-turbo";
        JSONArray messages = new JSONArray();
        int max_tokens = 4000;

        String prompt = String.format("Translate below into %s: %s", lvTargetLang, text);
        String escapedPrompt = prompt.replaceAll("\"", "\\\\\"").replaceAll("\n", "\\\\n");

        messages.put(new JSONObject().put("role", "system").put("content", "You are a translator."));
        messages.put(new JSONObject().put("role", "user").put("content", escapedPrompt));
        
        Map<String, String> headers = new TreeMap<>();
        headers.put("Content-Type", "application/json");
        headers.put("Authorization", "Bearer " + API_KEY);

        String body = new JSONObject()
                .put("model", model)
                .put("messages", messages)
                .put("max_tokens", max_tokens)
                .toString();

        System.out.println(body);
        String response = WikiGet.postJSON(API_URL, body, headers);
        JSONObject jsonResponse = new JSONObject(response);
        // Extract translated text to fit the new API response structure
        JSONArray choices = jsonResponse.getJSONArray("choices");
        if (choices.length() > 0) {
            JSONObject choice = choices.getJSONObject(0);
            if (choice.has("message")) {
            	JSONObject message = choice.getJSONObject("message");
                String translatedText = message.getString("content").trim();
                translationCache.put(cacheKey, translatedText);
                return translatedText;
            }
        }
        return "Translation failed";
    }
}
```

**Comment by ychoi-kr** [comment]
I opened new issue #3 .
